components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java

```diff
@@ -591,11 +593,35 @@ private void addOrUpdateOIDCSpProperty(OAuthAppDO oauthAppDO,
         PreparedStatement prepStatementForPropertyDelete =
                 connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries.REMOVE_SP_OIDC_PROPERTY);
 
-        if (isOIDCAudienceEnabled()) {
-            String[] audiences = oauthAppDO.getAudiences();
+        if (isOIDCIdTokenAudienceEnabled()) {
+            String[] audiences = oauthAppDO.getIdTokenAudiences();
+            HashSet<String> newAudiences = audiences == null ? new HashSet<>() : new HashSet<>(Arrays.asList
+                    (audiences));
+            List<String> oidcAudienceList = getOIDCIdTokenAudiences(spTenantDomain, oauthAppDO.getOauthConsumerKey());
+            Set<String> currentAudiences = oidcAudienceList == null ? new HashSet<>() : new HashSet<>(oidcAudienceList);
+            HashSet<String> newAudienceClone = (HashSet<String>) newAudiences.clone();
+            //removing all duplicate audiences in the new audience list
+            newAudiences.removeAll(currentAudiences);
+            //obtaining the audience values deleted in the list by user
+            currentAudiences.removeAll(newAudienceClone);
+
+            for (String deletedAudience : currentAudiences) {
+                addToBatchForOIDCPropertyDelete(preprocessedClientId, spTenantId, prepStatementForPropertyDelete,
+                        OPENID_CONNECT_ID_TOKEN_AUDIENCE, deletedAudience);
+            }
+
+            for (String addedAudience : newAudiences) {
+                addToBatchForOIDCPropertyAdd(preprocessedClientId, spTenantId, prepStatementForPropertyAdd,
+                        OPENID_CONNECT_ID_TOKEN_AUDIENCE, addedAudience);
+            }
+        }
```

Comment:
let's refactor the code to avoid code duplication with the below if block

Reply:
Resolved with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1902/commits/969e3978becce615f911d41d1b72521ce2bde2b2